File: JavaScript/searchRotatedSortedArray.js
const searchArray = (arr, target) => {
	let left = 0;
	let right = arr.length - 1;

	while (left <= right) {
		let mid = Math.floor((left + right) / 2);

		if (arr[mid] === target) {
			return mid;
		}

		// check which side of array is sorted
		if (arr[left] <= arr[mid]) {
			// if left side sorted
			if (arr[left] <= target && target <= arr[mid]) {
				right = mid - 1;
			} else {
				left = mid + 1;
			}
			// if right side sorted
		} else {
			if (arr[mid] <= target && target <= arr[right]) {
				left = mid + 1;
			} else {
				right = mid - 1;
			}
		}
	}
	return -1;
};

File: python/search_rotated_sorted_array.py
def search_rotated(arr, target):
    
    left = 0
    right = len(arr) - 1

    while left <= right:
        mid = (left + right) // 2

        if arr[mid] == target:
            return mid

        if arr[left] <= arr[mid]:
            if arr[left] <= target <= arr[mid]:
                right = mid - 1
            else:
                left = mid + 1
        else:
            if arr[mid] <= target <= arr[right]:
                left = mid + 1
            else:
                right = mid - 1
    return -1




arr = [4,5,6,7,0,1,2]
target = 0

if search_rotated(arr, target) == 4:
    print("Pass")
else:
    print("Fail")